README.md
# mahjong-scorer

A Python scorer for Mahjong (HK rules).

Licensed under MIT No Attribution (MIT-0), see [LICENSE].


## Usage

```bash
$ path/to/mahjongscore.py [-h] [-v] scores.txt

Score some games of Mahjong (HK rules).

positional arguments:
  scores.txt     name of scores file; output written to `{scores.txt}.tsv`

options:
  -h, --help     show this help message and exit
  -v, --version  show program's version number and exit
```


## Example

For the input scores file [`examples/comprehensive.txt`]:

```
B=2     # one & two bucks (一二蚊)
M=6     # maximum 6 faan (上限六番)
R=half  # half responsibility (半銃)
S=half  # half-spicy rise (半辣上)

A B C D
- - - -  # (摸和) draw
- - - f  # (詐糊) D false-win
1 - - -  # (自摸) A self-drawn win (1 faan)
- 2 d -  # (打出) B win (2 faan) from discard by C
- - 6 D  # (包打出) C win (6 faan) by discard guaranteed by D
- - 6 S  # (包自摸) C win (6 faan) by self-draw guaranteed by D
```

the output is the TSV file [`examples/comprehensive.txt.tsv`]:

| name | game_count | win_count | win_fraction | blame_count | blame_fraction | net_score | net_score_per_game |
| - | - | - | - | - | - | - | - |
| C | 6 | 2 | 0.3333 | 1 | 0.1667 | 500 | 83.3333 |
| B | 6 | 1 | 0.1667 | 0 | 0 | 204 | 34 |
| A | 6 | 1 | 0.1667 | 0 | 0 | 200 | 33.3333 |
| D | 6 | 0 | 0 | 3 | 0.5 | -904 | -150.6667 |
| * | 24 | 4 | 0.1667 | 4 | 0.1667 | 0 | 0 |


## Scores file syntax

The scorer reads a plain-text file of Mahjong scores.
Each line must have one of the following forms:

| Form | Meaning |
| - | - |
| `<yyyy>-<mm>-<dd>` | a date |
| `B=<base>` | a declaration of [base points] (default 1) |
| `M=<faan>` | a declaration of [maximum faan] (default 13) |
| `R=half \| full` | a declaration of [responsibility] (default full) |
| `S=half \| spicy` | a declaration of [spiciness] (default half) |
| `<P1> <P2> <P3> <P4>` | a declaration of [player names] (no hashes, asterisks, leading hyphens, or leading digits) |
| `<R1> <R2> <R3> <R4>` | a declaration of [game results] |
| `# <comment>` | a comment, also allowed at the end of the forms above |

See the linked sections for details.


## Details

### Base points

Declared via `B=<base>`.

For a 0-faan or chicken win (雞糊) under half responsibility (半銃):
- The two blameless players each pay `<base>/2` to the winner.
- The discarding (打出) player pays `<base>` to the winner.

| `<base>` | Cantonese name | English translation | Blameless payment | Discarding payment |
| - | - | - | - | - |
| 0.5 | 二五雞 | two & five chicken | 0.25 | 0.5 |
| 1 | 五一 | five & one | 0.5 | 1 |
| 2 | 一二蚊 | one & two bucks | 1 | 2 |

Base points are converted by a [multiplier] to a [portion].

### Maximum faan

Declared via `M=<faan>`.

In the event of a [false-win (詐糊)], the player at fault pays out the maximum
self-drawn (自摸) win amount to each other player.
This is to discourage collusion and deliberate false-winning.

### Responsibility

Declared via `R=half | full`.

| Responsibility | Cantonese name |
| - | - |
| `half` | 半銃 |
| `full` | 全銃 |

Determines whether a [discarding (打出)] player bears the payment of the two blameless players.

### Spiciness

Declared via `S=half | spicy`.

| Spiciness | Cantonese name | English translation |
| - | - | - |
| `half` | 半辣上 | half-spicy rise |
| `spicy` | 辣辣上 | spicy-spicy rise |

Determines whether the exponential rise in [multiplier] (with respect to number of faan)
is slowed to half-pace (with arithmetic-mean interpolation) when the number of faan exceeds 4.

#### Multiplier

| Number of faan | Multiplier for half-spicy rise<br>(半辣上) | Multiplier for spicy-spicy rise<br>(辣辣上) |
| - | - | - |
| 0 | 1 | 1 |
| 1 | 2 | 2 |
| 2 | 4 | 4 |
| 3 | 8 | 8 |
| 4 | 16 | 16 |
| 5 | 24 | 32 |
| 6 | 32 | 64 |
| 7 | 48 | 128 |
| 8 | 64 | 256 |
| 9 | 96 | 512 |
| 10 | 128 | 1024 |
| 11 | 192 | 2048 |
| 12 | 256 | 4096 |
| 13 | 384 | 8192 |

#### Portion

The [multiplier] is what converts [base points] to a portion:

$$ \text{Portion} = \text{Base Points} \times \text{Multiplier}. $$

### Player names

Declared via `<P1> <P2> <P3> <P4>`.

Each name must not have hashes, asterisks, leading hyphens, or leading digits.

### Game results

Declared via `<R1> <R2> <R3> <R4>`.

| `<Ri>` | Meaning |
| - | - |
| `<digits>` | winner's number of faan |
| `-` | blameless player |
| `d` | [discarding (打出)] player |
| `D` | [discard-guaranteeing (包打出)] player |
| `S` | [self-draw-guaranteeing (包自摸)] player |
| `f` | [false-winning (詐糊)] player |

Examples:

#### Draw (摸和)

```
A B C D
- - - -  # draw between A, B, C, D
```

Scores do not change (although the game count is incremented).

#### False-win (詐糊)

```
A B C D
- - - f  # D has a false-win
```

The player at fault pays out the maximum
self-drawn (自摸) win amount to each other player:
- Blameless players (A, B, C) each gain 3 [portions] at [maximum faan]
- Player at fault (D) loses 9 [portions] at [maximum faan]

#### Self-drawn win (自摸)

```
A B C D
8 - - -  # A has a self-drawn win (8 faan)
```

The blameless players each pay the winner one portion:
- Winner (A) gains 3 [portions] at the winning number of faan
- Blameless players (B, C, D) each lose 1 [portion] at the winning number of faan

#### Discarding (打出)

```
A B C D
8 - - d  # A wins (8 faan) from a discard by D
```

Under half [responsibility] (半銃):
- Winner (A) gains 2 [portions] at the winning number of faan
- Blameless players (B, C) each lose a 1/2 [portion] at the winning number of faan
- Player at fault (D) loses 1 [portion] at the winning number of faan

Under full [responsibility] (全銃):
- Winner (A) gains 2 [portions] at the winning number of faan
- Player at fault (D) loses 2 [portions] at the winning number of faan

#### Discard-guaranteeing (包打出)

To be used under half [responsibility] (半銃); redundant under full [responsibility] (全銃).

```
A B C D
8 - - D  # A wins (8 faan) by discard guaranteed by D
```

The player at fault bears all the losses of a discard win:
- Winner (A) gains 2 [portions] at the winning number of faan
- Player at fault (D) loses 2 [portions] at the winning number of faan

The envisaged situation is that A has been supplied enough favourable tiles
to have revealed melds indicating significant chance of a high-scoring win
(e.g. 12 same-suited tiles, 2 out of 3 dragon triplets, etc.).
Should such high-scoring win eventuate through discard by D,
then D ought to bear losses on behalf of B and C.

(Note that, if you are really that concerned about blameless players having to pay,
then you should consider playing under full [responsibility] (全銃) instead.)

#### Self-draw-guaranteeing (包自摸)

```
A B C D
8 - - S  # A wins (8 faan) by self-draw guaranteed by D
```

The player at fault bears all the losses of a self-drawn win:
- Winner (A) gains 3 [portions] at the winning number of faan
- Player at fault (D) loses 3 [portions] at the winning number of faan

The envisaged situation here is that D has supplied enough favourable tiles
for A to have revealed melds indicating significant chance of a high-scoring win
(e.g. 12 same-suited tiles, 2 out of 3 dragon triplets, etc.).
Should such high-scoring win eventuate through self-drawing by A,
then D ought to bear losses on behalf of B and C.


## Output data

| Column | Meaning |
| - | - |
| `name` | player name; bottom row `*` is the aggregate of all players |
| `game_count` | total number of games played |
| `win_count` | total number of games won |
| `win_fraction` | `win_count` divided by `game_count` |
| `blame_count` | total number of games to blame for |
| `blame_fraction` | `blame_count` divided by `game_count` |
| `net_score` | cumulative zero-sum score |
| `net_score_per_game` | `net_score` divided by `game_count` |


[LICENSE]: LICENSE
[`examples/comprehensive.txt`]: examples/comprehensive.txt
[`examples/comprehensive.txt.tsv`]: examples/comprehensive.txt.tsv

[base points]: #base-points
[maximum faan]: #maximum-faan
[responsibility]: #responsibility
[spiciness]: #spiciness
[multiplier]: #multiplier
[portion]: #portion
[portions]: #portion
[player names]: #player-names
[game results]: #game-results
[false-win (詐糊)]: #false-win-詐糊
[false-winning (詐糊)]: #false-win-詐糊
[discarding (打出)]: #discarding-打出
[discard-guaranteeing (包打出)]: #discard-guaranteeing-包打出
[self-draw-guaranteeing (包自摸)]: #self-draw-guaranteeing-包自摸


CHANGELOG.md
# Changelog


## [Unreleased]


## [v0.3.0] Date filters (2023-08-20)

- Implemented check for bad chronological order
- Implemented `--from yyyy-mm-dd` and `--to yyyy-mm-dd` date filters


## [v0.2.0] Blame (2023-07-09)

- Added blame count and blame fraction


## [v0.1.2] Indeterminate repair (2023-07-07)

- Fixed bug when sorting player with zero games
- Made indeterminate display as 'nan' in TSV


## [v0.1.1] No percent string formatting (2023-07-06)

- Rewrote `blunt` formatting without percent operator


## [v0.1.0] First acceptable (2023-06-13)

- First acceptable version.


[Unreleased]: https://github.com/yawnoc/mahjong-scorer/compare/v0.3.0...HEAD
[v0.3.0]: https://github.com/yawnoc/mahjong-scorer/compare/v0.2.0...v0.3.0
[v0.2.0]: https://github.com/yawnoc/mahjong-scorer/compare/v0.1.2...v0.2.0
[v0.1.2]: https://github.com/yawnoc/mahjong-scorer/compare/v0.1.1...v0.1.2
[v0.1.1]: https://github.com/yawnoc/mahjong-scorer/compare/v0.1.0...v0.1.1
[v0.1.0]: https://github.com/yawnoc/mahjong-scorer/releases/tag/v0.1.0


mahjong.py
#!/usr/bin/env python3

"""
# mahjong.py

A Python scorer for Mahjong (HK rules).

Copyright 2023 Conway
Licensed under MIT No Attribution (MIT-0), see LICENSE.
"""

import argparse
import csv
import os
import re
import sys

__version__ = '0.0.0'
DEFAULT_BASE = 1
DEFAULT_MAXIMUM_FAAN = 13
DEFAULT_RESPONSIBILITY = 'full'
DEFAULT_SPICINESS = 'half'


def get_duplicates(iterable):
    seen_items = set()
    duplicate_items = []
    for item in iterable:
        if item in seen_items:
            duplicate_items.append(item)
        else:
            seen_items.add(item)

    return duplicate_items


def robust_divide(dividend, divisor):
    try:
        return dividend / divisor
    except ZeroDivisionError:
        return None


def blunt(number, max_decimal_places=4):
    """
    Round a number to at most certain decimal places, as a string.
    """
    if number is None:
        return None

    if number == 0:
        return '0'

    max_decimal_places_format = f'%.{max_decimal_places}F'
    nice_string = max_decimal_places_format % number
    nice_string = re.sub(r'[.]?0*$', '', nice_string)

    return nice_string


class ScoreMaster:
    def __init__(self, scores_text):
        self.players_including_everyone, self.games = ScoreMaster.parse(scores_text)

    LINE_EXPLAINER = (
        f'A line must have one of the following forms:\n'
        f'    <yyyy>-<mm>-<dd>     # a date\n'
        f'    B=<base>             # a declaration of base points (default {DEFAULT_BASE}),\n'
        f'                         #   0.5 = two & five chicken (二五雞)\n'
        f'                         #   1   = five & one (五一)\n'
        f'                         #   2   = one & two bucks (一二蚊)\n'
        f'    M=<faan>             # a declaration of maximum faan (default {DEFAULT_MAXIMUM_FAAN})\n'
        f'    R=half | full        # a declaration of responsibility (default {DEFAULT_RESPONSIBILITY})\n'
        f'                         #   half  = half responsibility (半銃)\n'
        f'                         #   full  = full responsibility (全銃)\n'
        f'    S=half | spicy       # a declaration of spiciness (default {DEFAULT_SPICINESS})\n'
        f'                         #   half  = half-spicy rise (半辣上)\n'
        f'                         #   spicy = spicy-spicy rise (辣辣上)\n'
        f'    <P1> <P2> <P3> <P4>  # a declaration of player names (no hashes,\n'
        f'                         # asterisks, leading hyphens, or leading digits)\n'
        f'    <R1> <R2> <R3> <R4>  # a declaration of game results\n'
        f"                         #   <digits> = winner's faan\n"
        f'                         #   -        = blameless player\n'
        f'                         #   d        = discarding (打出) player\n'
        f'                         #   D        = discard-guaranteeing (包打出) player\n'
        f'                         #   S        = self-draw-guaranteeing (包自摸) player\n'
        f'                         #   f        = false-winning (詐糊) player\n'
        f'    # <comment>          # a comment, also allowed at the end of the forms\n'
        f'                         # above\n'
        f'All other lines are invalid.\n'
    )

    @staticmethod
    def parse(scores_text):
        player_from_name = {}
        games = []

        date = None
        base = DEFAULT_BASE
        maximum_faan = DEFAULT_MAXIMUM_FAAN
        responsibility = DEFAULT_RESPONSIBILITY
        spiciness = DEFAULT_SPICINESS
        names = None

        lines = scores_text.splitlines()
        for line_number, line in enumerate(lines, start=1):

            date_line_match = ScoreMaster.match_date_line(line)
            if date_line_match:
                date = date_line_match.group('date')
                continue

            base_line_match = ScoreMaster.match_base_line(line)
            if base_line_match:
                base_str = base_line_match.group('base')
                try:
                    base = float(base_str)
                except ValueError:
                    raise ScoreMaster.BadFloatException(
                        line_number,
                        f'unable to convert {base_str} to float',
                    )
                continue

            maximum_line_match = ScoreMaster.match_maximum_line(line)
            if maximum_line_match:
                maximum_faan = int(maximum_line_match.group('maximum_faan'))
                continue

            responsibility_line_match = ScoreMaster.match_responsibility_line(line)
            if responsibility_line_match:
                responsibility = responsibility_line_match.group('responsibility')
                continue

            spiciness_line_match = ScoreMaster.match_spiciness_line(line)
            if spiciness_line_match:
                spiciness = spiciness_line_match.group('spiciness')
                continue

            players_line_match = ScoreMaster.match_players_line(line)
            if players_line_match:
                names = tuple(
                    players_line_match.group(f'name_{i}')
                    for i in range(0, 4)
                )

                duplicate_names = get_duplicates(names)
                if duplicate_names:
                    raise ScoreMaster.DuplicatePlayerNamesException(
                        line_number,
                        f'duplicate player names {duplicate_names}',
                    )

                for name in names:
                    if name not in player_from_name:
                        player_from_name[name] = Player(name)
                continue

            game_line_match = ScoreMaster.match_game_line(line)
            if game_line_match:
                if names is None:
                    raise ScoreMaster.NoPlayersException(
                        line_number,
                        f'game declared without first declaring player names',
                    )

                faans = tuple(
                    ScoreMaster.normalise_faan(game_line_match.group(f'faan_{i}'))
                    for i in range(0, 4)
                )
                winner_index, winner_faan = ScoreMaster.extract_faan(faans, maximum_faan, line_number)

                blames = tuple(
                    ScoreMaster.normalise_blame(game_line_match.group(f'blame_{i}'))
                    for i in range(0, 4)
                )
                blame_index, blame_type = ScoreMaster.extract_blame(blames, line_number)

                if winner_index is None:
                    if blame_type is not None and blame_type != 'f':
                        raise ScoreMaster.NoWinYetNonFalseBlameException(
                            line_number,
                            f'game declared with no winner yet non-false-win blame (suffix `d`, `D`, or `S`)',
                        )
                else:
                    if blame_type == 'f':
                        raise ScoreMaster.WinYetFalseBlameException(
                            line_number,
                            f'game declared with winner yet false-win blame (suffix `f`)',
                        )

                if responsibility == 'full' and blame_type == 'D':
                    raise ScoreMaster.RedundantDiscardGuaranteeException(
                        line_number,
                        f'discard-guarantee `D` is redundant under full responsibility (全銃)',
                    )

                games.append(
                    Game(
                        date, base, maximum_faan, responsibility, spiciness,
                        names, winner_index, winner_faan, blame_index, blame_type,
                    )
                )
                continue

            if ScoreMaster.match_comment_line(line):
                continue

            raise ScoreMaster.InvalidLineException(
                line_number,
                f'invalid line\n\n{ScoreMaster.LINE_EXPLAINER}',
            )

        for game in games:
            game.update(player_from_name)

        players = list(player_from_name.values())
        everyone = Player('*')
        everyone.game_count = sum(p.game_count for p in players)
        everyone.win_count = sum(p.win_count for p in players)
        everyone.net_score = sum(p.net_score for p in players)

        players_including_everyone = players + [everyone]

        for player in players_including_everyone:
            player.update_averages()

        return players_including_everyone, games

    @staticmethod
    def match_date_line(line):
        return re.fullmatch(
            pattern=r'''
                ^ [\s]*
                (?P<date> [0-9]{4} - [0-9]{2} - [0-9]{2} )
                [\s]* (?: [#] .* )? $
            ''',
            string=line,
            flags=re.VERBOSE,
        )

    @staticmethod
    def match_base_line(line):
        return re.fullmatch(
            pattern=fr'''
                ^ [\s]*
                B=(?P<base> [.0-9]+ )
                [\s]* (?: [#] .* )? $
            ''',
            string=line,
            flags=re.VERBOSE,
        )

    @staticmethod
    def match_maximum_line(line):
        return re.fullmatch(
            pattern=fr'''
                ^ [\s]*
                M=(?P<maximum_faan> [0-9]+ )
                [\s]* (?: [#] .* )? $
            ''',
            string=line,
            flags=re.VERBOSE,
        )

    @staticmethod
    def match_responsibility_line(line):
        return re.fullmatch(
            pattern=fr'''
                ^ [\s]*
                R=(?P<responsibility> half | full )
                [\s]* (?: [#] .* )? $
            ''',
            string=line,
            flags=re.VERBOSE,
        )

    @staticmethod
    def match_spiciness_line(line):
        return re.fullmatch(
            pattern=fr'''
                ^ [\s]*
                S=(?P<spiciness> half | spicy )
                [\s]* (?: [#] .* )? $
            ''',
            string=line,
            flags=re.VERBOSE,
        )

    @staticmethod
    def match_players_line(line):
        player_name_regex = r'[^\s#*0-9-][^\s#*]*'
        return re.fullmatch(
            pattern=fr'''
                ^ [\s]*
                (?P<name_0> {player_name_regex} )
                    [\s]+
                (?P<name_1> {player_name_regex} )
                    [\s]+
                (?P<name_2> {player_name_regex} )
                    [\s]+
                (?P<name_3> {player_name_regex} )
                [\s]* (?: [#] .* )? $
            ''',
            string=line,
            flags=re.VERBOSE,
        )

    @staticmethod
    def match_game_line(line):
        faan_regex = '[0-9]+'
        blame_regex = '[-dDSf]'  # null, discard, discard-guarantee, self-draw-guarantee, or false-win
        return re.fullmatch(
            pattern=fr'''
                ^ [\s]*
                (?: (?P<faan_0> {faan_regex} ) | (?P<blame_0> {blame_regex} )  )
                    [\s]+
                (?: (?P<faan_1> {faan_regex} ) | (?P<blame_1> {blame_regex} )  )
                    [\s]+
                (?: (?P<faan_2> {faan_regex} ) | (?P<blame_2> {blame_regex} )  )
                    [\s]+
                (?: (?P<faan_3> {faan_regex} ) | (?P<blame_3> {blame_regex} )  )
                [\s]* (?: [#] .* )? $
            ''',
            string=line,
            flags=re.VERBOSE,
        )

    @staticmethod
    def match_comment_line(line):
        return re.fullmatch(
            pattern=r'^ [\s]* (?: [#] .* )? $',
            string=line,
            flags=re.VERBOSE,
        )

    @staticmethod
    def normalise_faan(faan_string):
        try:
            return int(faan_string)
        except TypeError:
            return None

    @staticmethod
    def normalise_blame(blame_string):
        if blame_string is None:
            return None

        if blame_string == '-':
            return None

        return blame_string

    @staticmethod
    def extract_faan(faans, maximum_faan, line_number):
        faan_indices = set(i for i in range(0, 4) if faans[i] is not None)

        if len(faan_indices) > 1:
            raise ScoreMaster.MultipleWinnersException(
                line_number,
                f'game declared with multiple winners (digits entries)',
            )

        try:
            winner_index = faan_indices.pop()
            winner_faan = faans[winner_index]
        except KeyError:
            winner_index = None
            winner_faan = None

        if winner_faan is not None and winner_faan > maximum_faan:
            raise ScoreMaster.MaximumFaanExceededException(
                line_number,
                f"game declared with winner's faan exceeding maximum faan ({maximum_faan})",
            )

        return winner_index, winner_faan

    @staticmethod
    def extract_blame(blames, line_number):
        blame_indices = set(i for i in range(0, 4) if blames[i] is not None)

        if len(blame_indices) > 1:
            raise ScoreMaster.MultipleBlameException(
                line_number,
                f'game declared with multiple players blamed (suffix `d`, `S`, or `f`)',
            )

        try:
            blame_index = blame_indices.pop()
            blame_type = blames[blame_index]
        except KeyError:
            blame_index = None
            blame_type = None

        return blame_index, blame_type

    def write_tsv(self, file_name):
        with open(file_name, 'w', encoding='utf-8', newline='') as file:
            writer = csv.writer(file, delimiter='\t', lineterminator=os.linesep)
            writer.writerow([
                'name',
                'game_count',
                'win_count',
                'win_fraction',
                'net_score',
                'net_score_per_game',
            ])
            for player in sorted(
                self.players_including_everyone,
                key=lambda p: (p.name == '*', -p.net_score_per_game, p.name),
            ):
                writer.writerow([
                    player.name,
                    player.game_count,
                    player.win_count,
                    blunt(player.win_fraction),
                    blunt(player.net_score),
                    blunt(player.net_score_per_game),
                ])

    class BadLineException(Exception):
        def __init__(self, line_number, message):
            self.line_number = line_number
            self.message = message

    class BadFloatException(BadLineException):
        pass

    class DuplicatePlayerNamesException(BadLineException):
        pass

    class NoPlayersException(BadLineException):
        pass

    class InvalidLineException(BadLineException):
        pass

    class MultipleWinnersException(BadLineException):
        pass

    class MaximumFaanExceededException(BadLineException):
        pass

    class MultipleBlameException(BadLineException):
        pass

    class NoWinYetNonFalseBlameException(BadLineException):
        pass

    class WinYetFalseBlameException(BadLineException):
        pass

    class RedundantDiscardGuaranteeException(BadLineException):
        pass


class Player:
    def __init__(self, name):
        self.name = name

        self.game_count = 0
        self.win_count = 0
        self.net_score = 0

        self.win_fraction = 0
        self.net_score_per_game = 0

    def update_averages(self):
        self.win_fraction = robust_divide(self.win_count, self.game_count)
        self.net_score_per_game = robust_divide(self.net_score, self.game_count)


class Game:
    def __init__(self, date, base, maximum_faan, responsibility, spiciness,
                 names, winner_index, winner_faan, blame_index, blame_type):
        self.date = date

        self.base = base
        self.maximum_faan = maximum_faan
        self.responsibility = responsibility
        self.spiciness = spiciness

        self.names = names
        self.winner_index = winner_index
        self.winner_faan = winner_faan
        self.blame_index = blame_index
        self.blame_type = blame_type

    def update(self, player_from_name):
        net_scores = Game.compute_net_scores(
            self.base, self.maximum_faan, self.responsibility, self.spiciness,
            self.winner_index, self.winner_faan, self.blame_index, self.blame_type,
        )

        for index, name in enumerate(self.names):
            player = player_from_name[name]
            player.game_count += 1
            player.win_count += 1 if index == self.winner_index else 0
            player.net_score += net_scores[index]

    @staticmethod
    def compute_net_scores(
        base, maximum_faan, responsibility, spiciness,
        winner_index, winner_faan, blame_index, blame_type
    ):
        if winner_index is None:  # no win

            if blame_index is None:  # draw (摸和)
                # Scores do not change.
                return (0, 0, 0, 0)

            elif blame_type == 'f':  # false-win (詐糊)
                # Blamed player pays each other player the maximum self-drawn win (i.e. three portions).
                portion = Game.compute_score_portion(base, spiciness, faan=maximum_faan)
                return tuple(
                    (-9 * portion) if i == blame_index else
                    (+3 * portion)
                    for i in range(0, 4)
                )

            raise RuntimeError(
                'Implementation error: `ScoreMaster.NoWinYetNonFalseBlameException` ought to have been raised'
            )

        else:  # win

            portion = Game.compute_score_portion(base, spiciness, faan=winner_faan)

            if blame_index is None:  # self-drawn win (自摸)
                # Blameless players each pay winner one portion.
                return tuple(
                    (+3 * portion) if i == winner_index else
                    (-portion)
                    for i in range(0, 4)
                )

            elif blame_type == 'd':  # discarding (打出)

                if responsibility == 'half':  # half responsibility (半銃)
                    # Blamed player pays winner one portion; blameless players each pay winner a half portion.
                    return tuple(
                        (+2 * portion) if i == winner_index else
                        (-portion) if i == blame_index else
                        (-portion/2)
                        for i in range(0, 4)
                    )

                elif responsibility == 'full':  # full responsibility (全銃)
                    # Blamed player pays winner a double portion.
                    return tuple(
                        (+2 * portion) if i == winner_index else
                        (-2 * portion) if i == blame_index
                        else 0
                        for i in range(0, 4)
                    )

                raise RuntimeError(
                    'Implementation error: `responsibility` is neither `half` nor `full`'
                )

            elif blame_type == 'D':  # discard-guaranteeing (包打出)
                # Blamed player pays winner a double portion; same as full responsibility (全銃).
                return tuple(
                    (+2 * portion) if i == winner_index else
                    (-2 * portion) if i == blame_index
                    else 0
                    for i in range(0, 4)
                )

            elif blame_type == 'S':  # self-draw-guaranteeing (包自摸)
                # Blamed player pays winner three portions.
                return tuple(
                    (+3 * portion) if i == winner_index else
                    (-3 * portion) if i == blame_index
                    else 0
                    for i in range(0, 4)
                )

            raise RuntimeError(
                'Implementation error: `ScoreMaster.WinYetFalseBlameException` ought to have been raised'
            )

    @staticmethod
    def compute_score_portion(base, spiciness, faan):
        if spiciness == 'half':  # half-spicy rise (半辣上)
            if faan <= 4:
                multiplier = 2 ** faan
            else:
                index = 4 + (faan - 4) // 2
                if faan % 2 == 0:
                    multiplier = 2 ** index
                else:
                    multiplier = 2 ** index * 3 // 2

        elif spiciness == 'spicy':  # spicy-spicy rise (辣辣上)
            multiplier = 2 ** faan

        else:
            raise RuntimeError(
                'Implementation error: `spiciness` is neither `half` nor `spicy`'
            )

        return base * multiplier


def parse_command_line_arguments():
    argument_parser = argparse.ArgumentParser(
        description='Score some games of Mahjong (HK rules).'
    )
    argument_parser.add_argument(
        '-v', '--version',
        action='version',
        version=f'{argument_parser.prog} version {__version__}'
    )
    argument_parser.add_argument(
        'scores_file_name',
        help='name of scores file; output written to `{scores.txt}.tsv`',
        metavar='scores.txt',
    )

    return argument_parser.parse_args()


def read_scores_text(scores_file_name):
    if os.path.isdir(scores_file_name):
        print(
            f'Error: `{scores_file_name}` is a directory, not a file',
            file=sys.stderr,
        )
        sys.exit(1)

    try:
        with open(scores_file_name, 'r', encoding='utf-8') as scores_file:
            scores_text = scores_file.read()
    except FileNotFoundError:
        print(
            f'Error: file `{scores_file_name}` not found',
            file=sys.stderr,
        )
        sys.exit(1)

    return scores_text


def main():
    parsed_arguments = parse_command_line_arguments()
    scores_file_name = parsed_arguments.scores_file_name

    scores_text = read_scores_text(scores_file_name)
    try:
        score_master = ScoreMaster(scores_text)
    except ScoreMaster.BadLineException as exception:
        line_number = exception.line_number
        message = exception.message
        print(
            f'Error (`{scores_file_name}`, line {line_number}): {message}'
        )
        sys.exit(1)

    tsv_file_name = f'{scores_file_name}.tsv'
    score_master.write_tsv(tsv_file_name)


if __name__ == '__main__':
    main()
